### PR TITLE
Add Subject, Photostim ,Tracking, Behavior Trials to nwb file

### PR DESCRIPTION
**Pull request by davidgodinez**
- Remove the Ephys portion of the nwb export function. @ttngu207  will include this portion pending further testing. 
- Reformat docker-compose file inside workers directory.

### CODE REVIEW

**Review by ttngu207 — APPROVED** [review]

File: pipeline/export/nwb.py [hunk 1]
```diff
@@ -6,11 +6,15 @@
 from dateutil.tz import tzlocal
 import pynwb
 from pynwb import NWBFile, NWBHDF5IO
+from pynwb.file import Subject
+from pynwb.behavior import SpatialSeries, Position
```
Comment (ttngu207) [inline pipeline/export/nwb.py:10]: you probably don't need these two, as there's already `pynwb` imported. You can use `pynwb.file.Subject` or `pynwb.behavior.SpatialSeries`

Comment (davidgodinez) [reply pipeline/export/nwb.py:10]: I will adjust.

**Comment by davidgodinez** [comment]
hi @ttngu207 I have added a new part of the function - tracking. Feel free to review at your earliest convenience.

File: pipeline/export/nwb.py [hunk 2]
```diff
@@ -163,6 +162,52 @@ def datajoint_to_nwb(session_key):
         nwbfile.add_ogen_site(stim_site)
         stim_sites[photostim['photo_stim']] = stim_site 
 
+    # =============================== TRACKING ===============================
```
Comment (davidgodinez) [inline pipeline/export/nwb.py:165]: @ttngu207 The tracking part of the function is here.

File: pipeline/export/nwb.py [hunk 3]
```diff
@@ -42,93 +48,293 @@ def datajoint_to_nwb(session_key):
...
+                        & session_key).proj(event_start='trial_event_time', event_stop = '(trial_event_time+duration)',
+                            outcome='outcome',trial_instruction='trial_instruction')
+
+    if q_behavior_trials:
+        trial_event_id, event_starts, event_stops, powers, photo_stim = q_behavior_trials.fetch(
+            'trial_event_id','event_start', 'event_stop', 'outcome', 'trial_instruction', order_by='trial')
+ 
+        all_times = [event_starts[0]]
+
+        for time in all_session_times.fetch('duration'):
```
Comment (dimitri-yatsenko) [inline pipeline/export/nwb.py:316]: Do not assume fetch is sorted.

File: pipeline/export/nwb.py [hunk 4]
```diff
@@ -42,93 +48,293 @@ def datajoint_to_nwb(session_key):
...
+            all_times.append(time)
+            all_times_arr = np.cumsum(all_times)
```
Comment (dimitri-yatsenko) [inline pipeline/export/nwb.py:318]: This cumsum should be outside the loop.

File: pipeline/export/nwb.py [hunk 5]
```diff
@@ -42,93 +48,293 @@ def datajoint_to_nwb(session_key):
...
+
+        all_times_arr = np.delete(all_times_arr, -1)
+
+        trial_start_times = []
+
+        for idx, times in enumerate(all_times_arr):
+            for trial in trial_event_id:
+                if trial == idx:
+                    trial_start_times.append(times)
```
Comment (dimitri-yatsenko) [inline pipeline/export/nwb.py:327]: There should be a much better / clearer way of doing this.

**Review by dimitri-yatsenko — CHANGES_REQUESTED** [review]

File: pipeline/export/nwb.py [hunk 6]
```diff
@@ -237,15 +233,16 @@ def datajoint_to_nwb(session_key):
...
+        # Add new table columns to nwb trial-table
         for c in trial_columns.values():
             nwbfile.add_trial_column(**c)
 
-        # Add entry to the trial-table
+        # Add entries to the trial-table
         for trial in q_trial_aggr.fetch(as_dict=True):
             trial['start_time'], trial['stop_time'] = float(trial['start_time']), float(trial['stop_time'])
             [trial.pop(k) for k in skip_adding_columns]
             nwbfile.add_trial(**trial)
```
Comment (dimitri-yatsenko) [inline pipeline/export/nwb.py:260]:
```suggestion
            nwbfile.add_trial(**{k: v for k, v in trial.items() if k not in skip_adding_columns})
```

File: pipeline/export/nwb.py [hunk 7]
```diff
@@ -30,9 +32,13 @@ def datajoint_to_nwb(session_key):
                               & (experiment.BehaviorTrial & session_key)).fetch1(
         'task_protocol_description')
 
+    try:
+        session_descr = (experiment.SessionComment & session_key).fetch1('session_comment')
+    except:
```
[the remark this answers is not in the document]
Comment (davidgodinez) [reply pipeline/export/nwb.py:37]: Thank you for your feedback @dimitri-yatsenko.
The error would be `DataJointError: fetch1 should only return one tuple. 0 tuples were found`

I will import `from datajoint.errors import DataJointError`

 try:
    session_descr = (experiment.SessionComment & session_key).fetch1('session_comment')
except DataJointError:
    session_descr = ''

File: pipeline/export/nwb.py [hunk 8]
```diff
@@ -42,93 +48,293 @@ def datajoint_to_nwb(session_key):
...
-
-            for attr in list(unit.keys()):
-                if attr in units_omitted_attributes:
-                    unit.pop(attr)
-                elif unit[attr] is None:
-                    unit[attr] = np.nan
-
-            nwbfile.add_unit(**unit)
-            pass
+    # electrodes_query = lab.ProbeType.Electrode * lab.ElectrodeConfig.Electrode
```
[the remark this answers is not in the document]
Comment (davidgodinez) [reply pipeline/export/nwb.py:59]: I will remove this.

File: pipeline/export/nwb.py [hunk 9]
```diff
@@ -42,93 +48,293 @@ def datajoint_to_nwb(session_key):
...
+                                                                                            * experiment.TrialEvent)
+                                                                                            & tracking_device
+                                                                                            & session_key).fetch(
+                                                                                                                'trial_event_id', 'trial_event_time',f'{table[1]}_x', 
+                                                                                                                f'{table[1]}_y',f'{table[1]}_likelihood', 'tracking_samples',
+                                                                                                                'duration') 
+                    # calculating start time for each trial_event_id relative to session
+                    all_start_times = (experiment.TrialEvent & session_key).fetch('duration')
+                    time_1 = (experiment.TrialEvent & session_key).fetch('trial_event_time', limit=1)
```
[the remark this answers is not in the document]
Comment (davidgodinez) [reply pipeline/export/nwb.py:195]: I will update this

File: pipeline/export/nwb.py [hunk 10]
```diff
@@ -42,92 +49,230 @@ def datajoint_to_nwb(session_key):
...
+                                                                    * tracking.TrackingDevice.proj('sampling_rate') 
+                                                                    & session_key).fetch(f'{name}_x',f'{name}_y', 
+                                                                                            f'{name}_likelihood', 
+                                                                                            'tracking_samples','start_time',
+                                                                                            'sampling_rate',order_by='trial')
+
+                recording_times_length = samples/sampling_rate.astype(float)
+                unshifted_time_stamps = [np.linspace(0,x, samples[idx].astype(int)) for idx, x in enumerate(recording_times_length)]
+                filled_start_times = [np.full(len(unshifted_time_stamps[idx]),x) for idx, x in enumerate(start_time)]
+                shifted_time_stamps = [np.add(filled_start_times[x].astype(float),unshifted_time_stamps[x]) for x in range(len(filled_start_times))]
```
Comment (ttngu207) [inline pipeline/export/nwb.py:197]: There must be a simpler way to do this, suggesting:
```suggestion
                shifted_time_stamps = [np.arange(sample) / fs + trial_start_time for sample, fs, trial_start_time in zip(samples, sampling_rate, start_time)]
```

File: pipeline/export/nwb.py [hunk 11]
```diff
@@ -42,92 +49,230 @@ def datajoint_to_nwb(session_key):
...
+
+                behav_acq.create_timeseries(name=f'BehavioralTimeSeries_{name}_x_y_data',
```
Comment (ttngu207) [inline pipeline/export/nwb.py:199]:
```suggestion
                behav_acq.create_timeseries(name=f'BehavioralTimeSeries_{name}', 
```

File: pipeline/export/nwb.py [hunk 12]
```diff
@@ -42,92 +49,230 @@ def datajoint_to_nwb(session_key):
...
+
+                recording_times_length = samples/sampling_rate.astype(float)
+                unshifted_time_stamps = [np.linspace(0,x, samples[idx].astype(int)) for idx, x in enumerate(recording_times_length)]
+                filled_start_times = [np.full(len(unshifted_time_stamps[idx]),x) for idx, x in enumerate(start_time)]
+                shifted_time_stamps = [np.add(filled_start_times[x].astype(float),unshifted_time_stamps[x]) for x in range(len(filled_start_times))]
+
+                behav_acq.create_timeseries(name=f'BehavioralTimeSeries_{name}_x_y_data', 
+                                            data=np.vstack([np.hstack(x),np.hstack(y), np.hstack(likelihood)]),
+                                            timestamps=np.hstack(shifted_time_stamps),
+                                            description='video description',
```
Comment (ttngu207) [inline pipeline/export/nwb.py:202]: update `'video description'` with something else?

Comment (davidgodinez) [reply pipeline/export/nwb.py]: changed video description to `description=f'Time series for {name} position: x, y, likelihood'`

File: pipeline/export/nwb.py (hunk 5 above)
Comment (davidgodinez) [reply pipeline/export/nwb.py]: This ended up being removed.

File: pipeline/export/nwb.py (hunk 6 above)
Comment (davidgodinez) [reply pipeline/export/nwb.py]: Changed.

File: pipeline/export/nwb.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (davidgodinez) [reply pipeline/export/nwb.py]: Commented code removed.

**Review by ttngu207 — APPROVED** [review]

**Comment by ttngu207** [comment]
Merge for now, however this is still a work in progress

**Pull request merged**